cylc/flow/id_cli.py

```diff
@@ -0,0 +1,482 @@
+            elif is_relative:
+                # this is a relative reference => expand it using the context
+                # of the partial ID
+                tokens_list.append(Tokens(
+                    **{
+                        **partials,
+                        **tokens,
+                    },
+                ))
+                partials_expended = True
+            else:
+                # this is a fully expanded reference
+                tokens_list.append(tokens)
+                partials = None
+                partials_expended = False
+        else:
+            # there was no previous reference that a relative reference
+            # could apply to
+            if is_partial:
+                partials = tokens
+                partials_expended = False
+            elif is_relative:
+                # so a relative reference is an error
+                raise ValueError(
+                    'Relative reference must follow an incomplete one.'
+                    '\nE.G: workflow //cycle/task'
+                )
+            else:
+                tokens_list.append(tokens)
+
+    if partials and not partials_expended:
+        # if the last ID was a "partial" but not expanded add it to the list
+        tokens_list.append(tokens)
+
+    return tokens_list
+
+
+def parse_ids(*args, **kwargs):
+    return asyncio.run(parse_ids_async(*args, **kwargs))
+
+
+async def parse_ids_async(
+    *ids: str,
+    src: bool = False,
+    match_workflows: bool = False,
+    match_active: Optional[bool] = True,
+    infer_latest_runs: bool = True,
+    constraint: str = 'tasks',
+    max_workflows: Optional[int] = None,
+    max_tasks: Optional[int] = None,
+) -> Tuple[Dict[str, List[Tokens]], Any]:
+    """Parse IDs from the command line.
+
+    Args:
+        ids:
+            Collection of IDs to parse.
+        src:
+            If True then source workflows can be provided via an absolute
+            path or a relative path starting "./".
+            Infers max_workflows = 1.
+        match_workflows:
+            If True workflows can be globs.
+        match_active:
+            If match_workflows is True this determines the wokflow state
+            filter.
+
+            True - running & paused
+            False - stopped
+            None - any
+        infer_latest_runs:
+            If true infer the latest run for a workflow when applicable
+            (allows 'cylc play one' rather than 'cylc play one/run1').
+        constraint:
+            Constrain the types of objects the IDs should relate to.
+
+            workflows - only allow workflows.
+            tasks - require tasks to be defined.
+            mixed - permit tasks not to be defined.
+        max_workflows:
+            Specify the maximum number of workflows permitted to be specified
+            in the ids.
+        max_tasks:
+            Specify the maximum number of tasks permitted to be specified
+            in the ids.
+
```

Comment:
IMO a "returns" section would be useful.

Reply:
Done.